Zend/zend_vm_def.h

```diff
@@ -8862,25 +8862,28 @@ ZEND_VM_HANDLER(183, ZEND_BIND_STATIC, CV, UNUSED, REF)
 
 	SAVE_OPLINE();
 	if (opline->extended_value & ZEND_BIND_REF) {
-		if (Z_TYPE_P(value) == IS_CONSTANT_AST) {
-			if (UNEXPECTED(zval_update_constant_ex(value, EX(func)->op_array.scope) != SUCCESS)) {
-				HANDLE_EXCEPTION();
-			}
-		}
-
 		i_zval_ptr_dtor(variable_ptr);
 		if (UNEXPECTED(!Z_ISREF_P(value))) {
 			zend_reference *ref = (zend_reference*)emalloc(sizeof(zend_reference));
 			GC_SET_REFCOUNT(ref, 2);
 			GC_TYPE_INFO(ref) = GC_REFERENCE;
-			ZVAL_COPY_VALUE(&ref->val, value);
+			if (OP2_TYPE == IS_UNUSED) {
+				ZVAL_COPY_VALUE(&ref->val, value);
+			} else {
+				i_zval_ptr_dtor(value);
```

Comment:
nit, not certain: This might not be common in practice, especially within `UNEXPECTED`. I assume zval_ptr_dtor (not inline) would result in smaller code size. (The `i_zval_ptr_dtor(variable_ptr);` above would be called much more often making inlining worth it)

Reply:
Looking again, I'm not certain about calling the destructor before replacing the static variable - it might call the destructor multiple times. (I'm assuming the initial value can be set from closure use?)


1. Ensure that this is only called once (and removes IS_TYPE_UNINITIALIZED once) when the destructor of the original static variable was 
2. Make this easier to reason about (e.g. with WeakMap as well)

E.g. what does this test currently do in debug builds (I think it'd access memory for $var that was being freed, but not certain of this?) (could add more tests of closure use regardless)

Copying to a temporary value with ZVAL_COPY_VALUE (if refcounted, or unconditionally) and only calling the destructor after the static variable was set would be a way to handle it if this caused issues

```php
class OrigVar {
    public function __destruct() {
        echo "in destructor\n";
        $GLOBALS['example']();
    }
}
$var = [new stdClass(), new OrigVar()];
$example = function () use ($var) {
    var_dump($var);
    static $var = 123;
    echo "after\n";
    var_dump($var);
};
$example();
```

Reply:
Oh. That combination is already forbidden in this RFC, I forgot about that unit test of [Zend/tests/static_variables_closure_bind.phpt](https://github.com/php/php-src/pull/9301/files#diff-80d24aadf4c683a8b3757eb4f2652b4dad2a0bbdb8446d39c8fc81c04ab29753)

Reply:
You're right, in this case the `(i_)zval_ptr_dtor` was useless as the value should always be constant.